File: 7.stacks_and_queues/lesson_7_fish.py
def solution(A, B):
    s = len(A)
    fight = True
    current_fish_index = 0
    last_fish_index = s-1

    while(fight):
        fight = False
        while(current_fish_index != last_fish_index):
            #if they meet
            if(B[current_fish_index] == 1 and B[current_fish_index + 1] == 0):
                    fight = True
                    last_fish_index -= 1
                    #we measure their size
                    if(A[current_fish_index] > A[current_fish_index + 1]):
                        A.pop(current_fish_index + 1)
                        B.pop(current_fish_index + 1)
                    else:
                        A.pop(current_fish_index)
                        B.pop(current_fish_index)
                        break
            else:
                current_fish_index += 1
        if(fight == True):
            current_fish_index = 0
    
    return last_fish_index + 1

#this was my first try - 100% correctness however only 50% performance because it is O(n**n)

#solution with stack from https://programming-review.com/python/algorithms#fish, which is O(n)

def solution(a, b):
    l = 0 # left
    s = [] # stack
    for i in range(len(a)):
        if b[i]==1: # fish moving upwards
            s.append(a[i]) # to the stack
        else: # eat fish mowing downwards.
            while len(s)>0:
                if s[-1]<a[i]:
                    s.pop() # eat from the stack
                else:
                    break
            else:
                l+=1
    l+=len(s)
    return l

File: 7.stacks_and_queues/lesson_7_fish.js
function solution(A, B) {
    let fish = Array.from(Array(A.length), () => new Array(2));
    for(let i = 0; i < A.length; i++) {
        fish[i][0] = A[i]
        fish[i][1] = B[i]
    }

    for(let i = 0; i < A.length - 1; i++) {
        let fish1 = fish[i]
        let fish2 = fish[i+1]
        if(fish2 == undefined || fish1 == undefined) return fish.length
        if(fish1[1] == 1 && fish2[1] == 0) {
            if(fish1[0] > fish2[0]) {
                fish.splice(i + 1, 1)
            } else {
                fish.splice(i, 1)
                if(i > 0) {
                    i--
                }
            }
            i--
        }
    }
    
    return fish.length
}

// first solution is giving 75% at perf
// second solution - using stack

function solution(A, B) {
    let stack = new Array()
    for(let i = 0; i < A.length; i++) {
        if(stack.length == 0) {
            stack.push([A[i], B[i]])
        } else {
            let fish = stack[stack.length - 1]
            let insert = true
            while(fish[1] == 1 && B[i] == 0) {
                if(fish[0] < A[i]) {
                    stack.pop()
                    if(stack.length == 0) break
                    fish = stack[stack.length - 1]
                } else {
                    insert = false
                    break
                }
            } 
            if(insert) {
                stack.push([A[i], B[i]])
            }
        }
    }
    return stack.length
}